Modules/_zstd/zdict.c

```diff
@@ -0,0 +1,296 @@
+    if (self->dict_content == NULL) {
+        PyErr_SetString(PyExc_TypeError,
+                        "dict_content argument should be bytes-like object.");
+        return -1;
+    }
+
+    /* Both ordinary dictionary and "raw content" dictionary should
+       at least 8 bytes */
+    if (Py_SIZE(self->dict_content) < 8) {
+        PyErr_SetString(PyExc_ValueError,
+                        "Zstd dictionary content should at least 8 bytes.");
+        return -1;
+    }
+
+    /* Get dict_id, 0 means "raw content" dictionary. */
+    self->dict_id = ZSTD_getDictID_fromDict(PyBytes_AS_STRING(self->dict_content),
+                                            Py_SIZE(self->dict_content));
+
+    /* Check validity for ordinary dictionary */
+    if (!is_raw && self->dict_id == 0) {
+        char *msg = "The dict_content argument is not a valid zstd "
+                    "dictionary. The first 4 bytes of a valid zstd dictionary "
+                    "should be a magic number: b'\\x37\\xA4\\x30\\xEC'.\n"
+                    "If you are an advanced user, and can be sure that "
+                    "dict_content argument is a \"raw content\" zstd "
+                    "dictionary, set is_raw parameter to True.";
+        PyErr_SetString(PyExc_ValueError, msg);
+        return -1;
+    }
+
+    return 0;
+}
+
+
+/*[clinic input]
+_zstd.ZstdDict.__reduce__
+
+
+Intentionally not supporting pickle.
+[clinic start generated code]*/
+
+static PyObject *
+_zstd_ZstdDict___reduce___impl(ZstdDict *self)
+/*[clinic end generated code: output=5c9b8a3550429417 input=2d69a2b4775ff76d]*/
+{
+    PyErr_SetString(PyExc_TypeError,
+                    "ZstdDict object intentionally doesn't support pickle. If you need "
+                    "to save zstd dictionary to disk, please save .dict_content "
+                    "attribute, it's a bytes object. So that the zstd dictionary "
+                    "can be used with other programs.");
+    return NULL;
+}
+
+#define clinic_state() (get_zstd_state(type))
+#include "clinic/zdict.c.h"
+#undef clinic_state
+
+static PyMethodDef ZstdDict_methods[] = {
+    _ZSTD_ZSTDDICT___REDUCE___METHODDEF
+
+    {0}
+};
+
+PyDoc_STRVAR(ZstdDict_dictid_doc,
+"ID of zstd dictionary, a 32-bit unsigned int value.\n\n"
+"Non-zero means ordinary dictionary, was created by zstd functions, follow\n"
+"a specified format.\n\n"
+"0 means a \"raw content\" dictionary, free of any format restriction, used\n"
+"for advanced user.");
+
+PyDoc_STRVAR(ZstdDict_dictcontent_doc,
+"The content of zstd dictionary, a bytes object, it's the same as dict_content\n"
+"argument in ZstdDict.__init__() method. It can be used with other programs.");
+
+static PyObject *
+ZstdDict_str(ZstdDict *dict)
+{
+    char buf[64];
+    PyOS_snprintf(buf, sizeof(buf),
+                  "<ZstdDict dict_id=%u dict_size=%zd>",
+                  dict->dict_id, Py_SIZE(dict->dict_content));
+
+    return PyUnicode_FromString(buf);
```

Comment:
Use `PyUnicode_Format`

Reply:
I used `PyUnicode_FromFormat`, which I presume is what you meant. Thank you for recommending all of these C APIs. I don't have  as much experience working directly with the C API so it's really helpful to have someone point out cleaner/better ways of doing things! I presume a lot of these things are the way they are because the extension was written several years ago.